client/src/components/AdminNewsletter/AdminNewsletter.tsx

```diff
@@ -1,27 +1,278 @@
+      const activeEmails = subscribers
+        .filter(subscriber => subscriber.active)
+        .map(subscriber => subscriber.email)
+        .join(', ');
+      clipboard.copy(activeEmails);
+    } catch (error) {
+      console.error('Failed to fetch or copy subscribers: ', error);
+    }
+  }
+
   return (
     <>
       <h4>Newsletter Admin Settings</h4>
-      <button onClick={copySubscribersToClipboard}>Copy To Clipboard</button>
+      <Tooltip
+        label="Subscriber Emails copied!"
+        offset={5}
+        position="bottom"
+        radius="xl"
+        transitionProps={{ duration: 100, transition: 'slide-down' }}
+        opened={clipboard.copied}
+      >
+      <Button
+        variant="light"
+        rightSection={
+          clipboard.copied ? (
+            <IconCheck style={{ width: rem(20), height: rem(20) }} stroke={1.5} />
+          ) : (
+            <IconCopy style={{ width: rem(20), height: rem(20) }} stroke={1.5} />
+          )
+        }
+        radius="xl"
+        size="md"
+        styles={{
+          root: { paddingRight: rem(14), height: rem(48) },
+          section: { marginLeft: rem(22) },
+        }}
+        onClick={handleCopyClick}
+      >
+        Copy Subscribers to clipboard
+      </Button>
+    </Tooltip>
+    <ScrollArea>
+      <TextInput
+        placeholder="Search by any field"
+        mb="md"
+        leftSection={<IconSearch style={{ width: rem(16), height: rem(16) }} stroke={1.5} />}
+        value={search}
+        onChange={handleSearchChange}
+      />
+      <Table horizontalSpacing="md" verticalSpacing="xs" miw={700} layout="fixed">
+        <Table.Tbody>
+          <Table.Tr>
+            <Th
+              sorted={sortBy === 'id'}
+              reversed={reverseSortDirection}
+              onSort={() => setSorting('id')}
+            >
+              id
+            </Th>
+            <Th
+              sorted={sortBy === 'firstname'}
+              reversed={reverseSortDirection}
+              onSort={() => setSorting('firstname')}
+            >
+              First Name
+            </Th>
+            <Th
+              sorted={sortBy === 'lastname'}
+              reversed={reverseSortDirection}
+              onSort={() => setSorting('lastname')}
+            >
+              Last Name
+            </Th>
+            <Th
+              sorted={sortBy === 'email'}
+              reversed={reverseSortDirection}
+              onSort={() => setSorting('email')}
+            >
+              Email
+            </Th>
+            <Th
+              sorted={sortBy === 'active'}
+              reversed={reverseSortDirection}
+              onSort={() => setSorting('active')}
+            >
+              Active
+            </Th>
+          </Table.Tr>
+        </Table.Tbody>
+        <Table.Tbody>
+          {rows.length > 0 ? (
+            rows
+          ) : (
+            <Table.Tr>
+              <Table.Td colSpan={Object.keys(data[0]).length}>
```

Comment:
Think you need to create a useState or holder for "data" instead of just local-scoping it within your helper functions

```
src/components/AdminNewsletter/AdminNewsletter.tsx:266:46 - error TS2304: Cannot find name 'data'.

266               <Table.Td colSpan={Object.keys(data[0]).length}>
                                                 ~~~~
Found 1 error in src/components/AdminNewsletter/AdminNewsletter.tsx:266

```